digicampipe/io/zfits.py

```diff
@@ -35,115 +34,173 @@ def zfits_event_source(
-
-        n_events = event_stream.numrows
-    else:
-
-        n_events = max_events
-
-    for event_counter, event in tqdm(enumerate(event_stream), total=n_events,
-                                     desc='Events', leave=False):
-        if max_events is not None and event_counter > max_events:
-            break
-
-        data.r0.event_id = event_counter
-        data.r0.tels_with_data = [event.telescope_id, ]
-
-        # remove forbidden telescopes
-        if allowed_tels:
-            data.r0.tels_with_data = [
-                list(filter(lambda x: x in data.r0.tels_with_data, sublist))
-                for sublist in allowed_tels
-            ]
-
-        for tel_id in data.r0.tels_with_data:
-
-            if tel_id not in loaded_telescopes:
-                data.inst.num_channels[tel_id] = event.num_channels
-                data.inst.num_pixels[tel_id] = event.n_pixels
-                data.inst.geom[tel_id] = geometry
-                data.inst.cluster_matrix_7[tel_id] = cluster_7_matrix
-                data.inst.cluster_matrix_19[tel_id] = cluster_19_matrix
-                data.inst.patch_matrix[tel_id] = patch_matrix
-                data.inst.num_samples[tel_id] = event.num_samples
-                loaded_telescopes.append(tel_id)
-
-            r0 = data.r0.tel[tel_id]
-            r0.camera_event_number = event.event_number
-            r0.pixel_flags = event.pixel_flags
-            r0.local_camera_clock = event.local_time
-            r0.gps_time = event.central_event_gps_time
-            r0.camera_event_type = event.camera_event_type
-            r0.array_event_type = event.array_event_type
-            r0.adc_samples = event.adc_samples
-
-            r0.trigger_input_traces = event.trigger_input_traces
-            r0.trigger_output_patch7 = event.trigger_output_patch7
-            r0.trigger_output_patch19 = event.trigger_output_patch19
-            r0.digicam_baseline = event.baseline
-
-        yield data
-
-    return
+    with SimpleFile(url) as file:
+
+        for event_counter, event in tqdm(
+            enumerate(file.Events),
+            desc='Events',
+            leave=False
+        ):
+            if max_events is not None and event_counter > max_events:
+                break
+
+            data.r0.event_id = event_counter
+            data.r0.tels_with_data = [event.telescopeID, ]
+
+            # remove forbidden telescopes
+            if allowed_tels:
+                data.r0.tels_with_data = [
+                    list(
+                        filter(
+                            lambda x: x in data.r0.tels_with_data, sublist))
+                    for sublist in allowed_tels
+                ]
+
+            _sort_ids = None
+
+            for tel_id in data.r0.tels_with_data:
+
+                pixel_ids = event.hiGain.waveforms.pixelsIndices
+                n_pixels = len(pixel_ids)
+                if _sort_ids is None:
+                    _sort_ids = np.argsort(pixel_ids)
```

Comment:
I think such sorting should happen in the writer or even before. Main reason being you only what to do this once. But maybe it is for latter.